common/json_writer/src/lib.rs

```diff
@@ -22,46 +23,62 @@ pub enum JsonWriterError {
 
 impl JsonWriter {
     pub fn new() -> Self {
-        Self { buffer: Vec::new() }
+        Self {
+            buffer: Vec::new(),
+            needs_separator: false,
+        }
     }
 
     pub fn with_capacity(capacity: usize) -> Self {
         Self {
             buffer: Vec::with_capacity(capacity),
+            needs_separator: false,
         }
     }
 
     pub fn write_key(&mut self, key: &str) -> Result<(), JsonWriterError> {
         self.write_str(key)?;
```

Comment:
I would call `separate()` *before* writing the key.

Reply:
ah, that's tricky, because `self.write_str` is a public method that itself writes out a separator.

Reply:
Ok, I understand the logic now. I missed the fact that the key was written using `self.write_str()` hence calling the `separate()`. This way is better than what I had in mind because also working with arrays.